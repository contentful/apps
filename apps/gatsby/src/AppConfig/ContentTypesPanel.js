import {
  CheckboxField,
  FieldGroup,
  Heading,
  Note,
  Paragraph,
  SkeletonBodyText,
  SkeletonContainer,
  TextLink,
  Typography,
} from "@contentful/forma-36-react-components";
import styles from "../styles";
import React from "react";

const ContentTypesSkeleton = () => (
  <SkeletonContainer width="100%">
    <SkeletonBodyText numberOfLines={3} />
  </SkeletonContainer>
);

const NoContentTypes = ({ space, environment }) => (
  <Note noteType="warning">
    There are no content types available in this environment. You can add one{" "}
    <TextLink
      target="_blank"
      href={
        environment === "master"
          ? `https://app.contentful.com/spaces/${space}/content_types`
          : `https://app.contentful.com/spaces/${space}/environments/${environment}/content_types`
      }
      rel="noopener noreferrer"
    >
      content type
    </TextLink>{" "}
    and assign it to the app from this screen.
  </Note>
);

export const ContentTypesList = ({
  contentTypes,
  enabledContentTypes,
  onContentTypeToggle,
  space,
  environment,
}) => {
  if (!contentTypes) {
    return <ContentTypesSkeleton />;
  }

  if (0 === contentTypes.length) {
    return <NoContentTypes space={space} environment={environment} />;
  }

  return contentTypes.map(({ sys, name }) => (
    <div>
      <CheckboxField
        key={sys.id}
        labelIsLight
        labelText={name}
        name={name}
        checked={enabledContentTypes.includes(sys.id)}
        value={sys.id}
        onChange={() => onContentTypeToggle(sys.id)}
        id={sys.id}
      />
    </div>
  ));
};

const ContentTypesPanel = ({
  contentTypes,
  enabledContentTypes,
  onContentTypeToggle,
  space,
  environment,
}) => (
  <Typography>
    <Heading>Content Types</Heading>
    <Paragraph>
      Select content types that will show the Gatsby Cloud functionality in the
      sidebar.
    </Paragraph>
    <div className={styles.checks}>
      <FieldGroup>
        <ContentTypesList
          space={space}
          environment={environment}
          contentTypes={contentTypes}
          enabledContentTypes={enabledContentTypes}
          onContentTypeToggle={onContentTypeToggle}
        />
      </FieldGroup>
    </div>
  </Typography>
);

export default ContentTypesPanel;
