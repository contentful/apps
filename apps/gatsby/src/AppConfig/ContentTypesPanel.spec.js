import React from "react";
import { ContentTypesList } from "./ContentTypesPanel";
import { cleanup, render } from "@testing-library/react";

describe("ContentTypesList", function() {
  let contentTypeId = 1;
  const environment = "master";
  const space = "12512asfasf";

  const contentType = (name) => ({
    sys: { id: String(++contentTypeId) },
    name,
  });
  const queryByInputName = (name) => (_, element) =>
    element.tagName.toLowerCase() === "input" && element.name === name;

  afterEach(cleanup);

  it("should show Skeleton on nil values", () => {
    const { container } = render(<ContentTypesList contentTypes={null} />);

    expect(container).toMatchSnapshot();
  });

  it("should show a Note if there are no content types", () => {
    const { container } = render(
      <ContentTypesList
        contentTypes={[]}
        space={space}
        environment={environment}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("should show checkboxes per content type", () => {
    const contentTypes = [contentType("posts"), contentType("authors")];
    const { queryByText } = render(
      <ContentTypesList
        contentTypes={contentTypes}
        enabledContentTypes={[]}
        space={space}
        environment={environment}
      />
    );

    expect(queryByText("posts")).toBeDefined();
    expect(queryByText("authors")).toBeDefined();
  });

  it("should check enabled content types", () => {
    const posts = contentType("posts");
    const authors = contentType("authors");
    const contentTypes = [posts, authors];
    const enabledTypes = [posts.sys.id];
    const { queryByText } = render(
      <ContentTypesList
        contentTypes={contentTypes}
        enabledContentTypes={enabledTypes}
        environment={environment}
        space={space}
      />
    );

    const postsElement = queryByText(queryByInputName("posts"));
    const authorsElement = queryByText(queryByInputName("authors"));

    expect(postsElement).toBeChecked();
    expect(authorsElement).not.toBeChecked();
  });
});
