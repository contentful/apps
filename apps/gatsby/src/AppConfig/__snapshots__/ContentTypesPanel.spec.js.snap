// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentTypesList should show Skeleton on nil values 1`] = `
<div>
  <svg
    aria-label="Loading component..."
    class="SkeletonContainer__SkeletonContainer___23jiu"
    data-test-id="cf-ui-skeleton-form"
    height="100%"
    preserveAspectRatio="xMidYMid meet"
    role="img"
    width="100%"
  >
    <title>
      Loading component...
    </title>
    <rect
      clip-path="url(#cf-ui-skeleton-clip-0)"
      height="100%"
      style="fill: url(#cf-ui-skeleton-clip-gradient-1);"
      width="100%"
      x="0"
      y="0"
    />
    <defs>
      <clippath
        id="cf-ui-skeleton-clip-0"
      >
        <rect
          height="16"
          rx="0"
          ry="0"
          width="100%"
          x="0"
          y="0"
        />
        <rect
          height="16"
          rx="0"
          ry="0"
          width="100%"
          x="0"
          y="24"
        />
        <rect
          height="16"
          rx="0"
          ry="0"
          width="80%"
          x="0"
          y="48"
        />
      </clippath>
      <lineargradient
        id="cf-ui-skeleton-clip-gradient-1"
      >
        <stop
          offset="0%"
          stop-color="#e5ebed"
          stop-opacity="1"
        >
          <animate
            attributeName="stop-color"
            dur="2s"
            id="animation-2"
            repeatCount="indefinite"
            values="#e5ebed; #f7f9fa; #e5ebed"
          />
        </stop>
        <stop
          offset="50%"
          stop-color="#f7f9fa"
          stop-opacity="1"
        >
          <animate
            attributeName="stop-color"
            begin="animation-2.begin+0.25s"
            dur="2s"
            repeatCount="indefinite"
            values="#e5ebed; #f7f9fa; #e5ebed"
          />
        </stop>
        <stop
          offset="100%"
          stop-color="#e5ebed"
          stop-opacity="1"
        >
          <animate
            attributeName="stop-color"
            begin="animation-2.begin+0.5s"
            dur="2s"
            repeatCount="indefinite"
            values="#e5ebed; #f7f9fa; #e5ebed"
          />
        </stop>
      </lineargradient>
    </defs>
  </svg>
</div>
`;

exports[`ContentTypesList should show a Note if there are no content types 1`] = `
<div>
  <div
    class="Note__Note___2eSKN Note__Note--warning___3X53I"
    data-test-id="cf-ui-note"
  >
    <div
      class="Note__Note__icon___20RqC"
    >
      <svg
        class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--warning___39Bnz"
        data-test-id="cf-ui-icon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
        />
      </svg>
    </div>
    <div
      class="Note__Note__info___2m_-K"
    >
      <div
        class="Note__Note__content___H1pf-"
      >
        There are no content types available in this environment. You can add one
         
        <a
          class="TextLink__TextLink___1biUr a11y__focus-outline--default___2hwb1 TextLink__TextLink--primary___2Vc9F"
          data-test-id="cf-ui-text-link"
          href="https://app.contentful.com/spaces/12512asfasf/content_types"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            class="TabFocusTrap__TabFocusTrap___39Vty"
            tabindex="-1"
          >
            content type
          </span>
        </a>
         
        and assign it to the app from this screen.
      </div>
    </div>
  </div>
</div>
`;
